arch/nios2/core/timing.c

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (c) 2020 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <sys_clock.h>
+#include <timing/timing.h>
+#include "altera_avalon_timer_regs.h"
+
+#define NIOS2_SUBTRACT_CLOCK_CYCLES(val)                                       \
+	((IORD_ALTERA_AVALON_TIMER_PERIODH(TIMER_0_BASE) << 16 |               \
+	  (IORD_ALTERA_AVALON_TIMER_PERIODL(TIMER_0_BASE))) -                  \
+	 ((uint32_t)val))
+
+#define TIMING_INFO_OS_GET_TIME()                                              \
+	(NIOS2_SUBTRACT_CLOCK_CYCLES(                                          \
+		((uint32_t)IORD_ALTERA_AVALON_TIMER_SNAPH(TIMER_0_BASE)        \
+		 << 16) |                                                      \
+		((uint32_t)IORD_ALTERA_AVALON_TIMER_SNAPL(TIMER_0_BASE))))
+
+void __weak timing_init(void)
```

Comment:
what's the use-case for overriding these at the arch instead of common level?

Reply:
an architecture can either use the default and not provide its own or implement its own timing functions in a way that can be overridden by an SoC or a board or an application.

Reply:
I think I see. I'm not sure if `__weak` is the best way to approach this as it's not completely clear to me if this will work with multiple layers of overrides. We might have to use a vtable or something instead even though it's a bit of extra overhead if we want it to be completely flexible.